src/TSTransformer/nodes/expressions/transformUnaryExpression.ts

```diff
@@ -7,57 +7,74 @@ import { transformExpression } from "TSTransformer/nodes/expressions/transformEx
 import { transformWritableExpression } from "TSTransformer/nodes/transformWritable";
 import { createTruthinessChecks } from "TSTransformer/util/createTruthinessChecks";
 import { getKindName } from "TSTransformer/util/getKindName";
+import { isUsedAsStatement } from "TSTransformer/util/isUsedAsStatement";
 import { isDefinitelyType, isNumberType } from "TSTransformer/util/types";
-import { validateNotAnyType } from "TSTransformer/util/validateNotAny";
 import ts from "typescript";
 
+function checkUnaryType(state: TransformState, node: ts.PrefixUnaryExpression | ts.PostfixUnaryExpression): void {
+	// We can omit passing `node` because `any` won't match `number`
+	// The noNonNumberUnary error is more accurate than noAny
+	if (!isDefinitelyType(state, state.getType(node.operand), undefined, isNumberType)) {
+		DiagnosticService.addDiagnostic(errors.noNonNumberUnary(node));
+	}
+}
+
 export function transformPostfixUnaryExpression(state: TransformState, node: ts.PostfixUnaryExpression) {
-	validateNotAnyType(state, node.operand);
+	// TS will allow both `any` and `number`
+	checkUnaryType(state, node);
 
-	const writable = transformWritableExpression(state, node.operand, true);
-	const origValue = luau.tempId("original");
+	const originalIsUsed = !isUsedAsStatement(node);
 
-	state.prereq(
-		luau.create(luau.SyntaxKind.VariableDeclaration, {
-			left: origValue,
-			right: writable,
-		}),
-	);
-
-	const operator: luau.AssignmentOperator = node.operator === ts.SyntaxKind.PlusPlusToken ? "+=" : "-=";
+	const writable = transformWritableExpression(state, node.operand, originalIsUsed);
+	const origValue = luau.tempId("original");
+	if (originalIsUsed) {
+		state.prereq(
+			luau.create(luau.SyntaxKind.VariableDeclaration, {
+				left: origValue,
+				right: writable,
+			}),
+		);
+	}
 
 	state.prereq(
 		luau.create(luau.SyntaxKind.Assignment, {
 			left: writable,
-			operator,
+			operator:
+				node.operator === ts.SyntaxKind.PlusPlusToken
+					? "+="
+					: node.operator === ts.SyntaxKind.MinusMinusToken
+					? "-="
+					: assert(false, "Unknown postfix unary operator"),
 			right: luau.number(1),
 		}),
 	);
 
-	return origValue;
+	return originalIsUsed ? origValue : luau.nil();
 }
 
 export function transformPrefixUnaryExpression(state: TransformState, node: ts.PrefixUnaryExpression) {
-	validateNotAnyType(state, node.operand);
+	// All operators except `!` need operand to be `number` type
+	if (node.operator !== ts.SyntaxKind.ExclamationToken) checkUnaryType(state, node);
```

Comment:
Seems like a better idea to put this in the top of each case. TypeScript might add new unary expression types in the future.

Reply:
Better to be on the safe side for those new cases then, right? They'll assertion fail just past that anyways for not being handled.

Reply:
Either way we won't handle new cases correctly when they're first introduced. Putting it at the top of each case makes the intent more clear.

If someone is looking up "How does the X operator" work, they'll likely just read the if statement instead of the entire function.